File: fs/before_checksum/68__text-justification/cpp/text-justification_benchmark_v2.py
# Time:  O(n)
# Space: O(k), k is maxWidth.

class Solution(object):
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        def addSpaces(i, spaceCnt, maxWidth, is_last):
            if i < spaceCnt:
                # For the last line of text, it should be left justified,
                # and no extra space is inserted between words.
                return 1 if is_last else (maxWidth // spaceCnt) + int(i < maxWidth % spaceCnt)
            return 0

        def connect(words, maxWidth, begin, end, length, is_last):
            s = []  # The extra space O(k) is spent here.
            n = end - begin
            for i in xrange(n):
                s += words[begin + i],
                s += ' ' * addSpaces(i, n - 1, maxWidth - length, is_last),
            # For only one word in a line.
            line = "".join(s)
            if len(line) < maxWidth:
                line += ' ' * (maxWidth - len(line))
            return line

        res = []
        begin, length = 0, 0
        for i in xrange(len(words)):
            if length + len(words[i]) + (i - begin) > maxWidth:
                res += connect(words, maxWidth, begin, i, length, False),
                begin, length = i, 0
            length += len(words[i])

        # Last line.
        res += connect(words, maxWidth, begin, len(words), length, True),
        return res

if __name__ == "__main__":
    tests = [
        (["Hello"], 5),
        (["Hi"], 4),
        (["This","is","an","example","of","text","justify"], 16),
        (["a","b","c"], 6),
        (["lorem","ipsum","dolor","sit","amet","consect","adipis"], 14),
        (["a","group","of","words","to","fill","the","last","line"], 12),
        (["aa","bb","cc","dd"], 11),
        (["a","b","c","d"], 1),
        (["longword","tiny","mid","size","words","here"], 10),
        (["abc","def","ghi"], 11)
    ]

    sol = Solution()
    sink = 0

    iterations = 1
    for _ in xrange(iterations):
        checksum = 0
        for words, width in tests:
            res = sol.fullJustify(list(words), width)
            for line in res:
                checksum += len(line)
                checksum += ord(line[0])
                checksum += ord(line[-1])
            sink += checksum
    print(sink)

File: fs/cpp_before_energy/68__text-justification/text-justification_benchmark_v2.cpp
#include <iostream>
template <class T>
static inline void DoNotOptimize(const T& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

#include <vector>
#include <string>

class Solution {
private:
    static int addSpaces(int i, int spaceCnt, int extraWidth, bool is_last) {
        if (i < spaceCnt) {
            if (is_last) return 1;
            int base = spaceCnt == 0 ? 0 : (extraWidth / spaceCnt);
            int rem = spaceCnt == 0 ? 0 : (extraWidth % spaceCnt);
            return base + (i < rem ? 1 : 0);
        }
        return 0;
    }

    static std::string connect(const std::vector<std::string>& words, int maxWidth, int begin, int end, int length, bool is_last) {
        std::string line;
        int n = end - begin;
        for (int i = 0; i < n; ++i) {
            line += words[begin + i];
            int spaces = addSpaces(i, n - 1, maxWidth - length, is_last);
            if (spaces > 0) line.append(spaces, ' ');
        }
        if (static_cast<int>(line.size()) < maxWidth) {
            line.append(maxWidth - line.size(), ' ');
        }
        return line;
    }

public:
    std::vector<std::string> fullJustify(std::vector<std::string>& words, int maxWidth) {
        std::vector<std::string> res;
        int begin = 0;
        int length = 0;
        for (int i = 0; i < static_cast<int>(words.size()); ++i) {
            if (length + static_cast<int>(words[i].size()) + (i - begin) > maxWidth) {
                res.emplace_back(connect(words, maxWidth, begin, i, length, false));
                begin = i;
                length = 0;
            }
            length += static_cast<int>(words[i].size());
        }
        res.emplace_back(connect(words, maxWidth, begin, static_cast<int>(words.size()), length, true));
        return res;
    }
};

struct Case {
    std::vector<std::string> words;
    int width;
};

int main() {
    std::vector<Case> tests = {
        {{"Hello"}, 5},
        {{"Hi"}, 4},
        {{"This","is","an","example","of","text","justify"}, 16},
        {{"a","b","c"}, 6},
        {{"lorem","ipsum","dolor","sit","amet","consect","adipis"}, 14},
        {{"a","group","of","words","to","fill","the","last","line"}, 12},
        {{"aa","bb","cc","dd"}, 11},
        {{"a","b","c","d"}, 1},
        {{"longword","tiny","mid","size","words","here"}, 10},
        {{"abc","def","ghi"}, 11}
    };

    Solution sol;
    volatile int sink = 0;

    const int iterations = 1000;
    for (int iter = 0; iter < iterations; ++iter) {
        //int checksum = 0;
        for (const auto& tc : tests) {
            std::vector<std::string> words = tc.words;
            std::vector<std::string> res = sol.fullJustify(words, tc.width);
            DoNotOptimize(res); 
            // for (const auto& line : res) {
            //     checksum += static_cast<int>(line.size());
            //     checksum += static_cast<unsigned char>(line[0]);
            //     checksum += static_cast<unsigned char>(line[line.size() - 1]);
            // }
        }
        //sink = checksum;
    }

    return 0;
}